pkg/awsiamauth/reconciler/reconciler.go

```diff
@@ -0,0 +1,123 @@
+func (r *Reconciler) ReconcileAWSIAMAuthCASecret(ctx context.Context, logger logr.Logger, client client.Client, clusterName string) (controller.Result, error) {
+	// Fetch the CA Secret
+	s := &corev1.Secret{}
+	err := client.Get(ctx, types.NamespacedName{Name: awsiamauth.GetAwsIamAuthCaSecretName(clusterName), Namespace: constants.EksaSystemNamespace}, s)
+	if apierrors.IsNotFound(err) {
+		logger.Info("Applying aws-iam-authenticator CA secret")
+		return r.applyCASecret(ctx, client, clusterName)
+	}
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("fetching secret %s: %v", awsiamauth.GetAwsIamAuthCaSecretName(clusterName), err)
+	}
+
+	// If Secret is present, no-op
+	logger.Info("aws-iam-authenticator CA secret found", "Name", awsiamauth.GetAwsIamAuthCaSecretName(clusterName))
+	return controller.Result{}, nil
+}
+
+func (r *Reconciler) applyCASecret(ctx context.Context, client client.Client, clusterName string) (controller.Result, error) {
+	// Generate k8s secret yaml for aws-iam-authenticator
+	yaml, err := r.templateBuilder.GenerateCertKeyPairSecret(r.certgen, clusterName)
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("generating aws-iam-authenticator ca secret: %v", err)
+	}
+
+	return controller.Result{}, serverside.ReconcileYaml(ctx, client, yaml)
+}
+
+// Reconcile takes the AWS IAM Authenticator indentity provider to the desired state defined in AWSIAMConfig.
+// It uses a controller.Result to indicate when requeues are needed.
+// Intended to be used in a kubernetes controller.
+func (r *Reconciler) Reconcile(ctx context.Context, logger logr.Logger, client client.Client, cluster *anywherev1.Cluster) (controller.Result, error) {
+	// Build the cluster spec for the Cluster
+	clusterSpec, err := anywhereCluster.BuildSpec(ctx, clientutil.NewKubeClient(client), cluster)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	// Get the client object for the workload cluster
+	rClient, err := r.remoteClientRegistry.GetClient(ctx, controller.CapiClusterObjectKey(cluster))
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	var clusterID uuid.UUID
+	// Fetch the aws-iam-authenticator config map.
+	cm := &corev1.ConfigMap{}
+	err = rClient.Get(ctx, types.NamespacedName{Name: awsiamauth.AwsIamAuthConfigMapName, Namespace: constants.KubeSystemNamespace}, cm)
+	if apierrors.IsNotFound(err) {
+		// Use a new clusterID
+		clusterID = r.clusterID
+		return r.applyIAMAuthManifest(ctx, rClient, clusterSpec, clusterID)
+	}
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("fetching configmap %s: %v", awsiamauth.AwsIamAuthConfigMapName, err)
+	}
+
+	logger.Info("Applying aws-iam-authenticator manifest")
+	return r.applyIAMAuthManifest(ctx, rClient, clusterSpec, clusterID)
+}
+
+func (r *Reconciler) applyIAMAuthManifest(ctx context.Context, client client.Client, clusterSpec *anywhereCluster.Spec, clusterID uuid.UUID) (controller.Result, error) {
```

Comment:
We should probably document that `clusterID` can be not set and this method will respect the current value. Also, should we make it a pointer so it can actually be nil?

Reply:
Actually we check for UUID.Nil in the [GenerateManifest](https://github.com/aws/eks-anywhere/blob/main/pkg/awsiamauth/template.go#L22) function. Thats why i hadn't used pointers. I could probably look into refactoring that separately since it changes the CLI create as well.